src/main/java/com/iota/iri/service/garbageCollector/GarbageCollector.java

```diff
@@ -0,0 +1,323 @@
+    }
+
+    /**
+     * This method persists the changes of the garbage collector so IRI can continue cleaning up upon restarts.
+     *
+     * Since cleaning up the old database entries can take a long time, we need to make sure that it is possible to
+     * continue where we stopped. This method therefore writes the state of the queued jobs into a file that is read
+     * upon re-initialization of the GarbageCollector.
+     *
+     * @throws GarbageCollectorException if something goes wrong while writing the state file
+     */
+    void persistChanges() throws GarbageCollectorException {
+        try {
+            Files.write(
+                Paths.get(getStateFile().getAbsolutePath()),
+                () -> garbageCollectorJobs.entrySet().stream().flatMap(
+                    entry -> entry.getValue().stream()
+                ).<CharSequence>map(
+                    entry -> entry.getClass().getCanonicalName() + ";" + entry.serialize()
+                ).iterator()
+            );
+        } catch(IOException e) {
+            throw new GarbageCollectorException("could not persists garbage collector state", e);
+        }
+    }
+
+    /**
+     * This method contains the logic for scheduling the jobs and executing them.
+     *
+     * It iterates through all available queues and executes the corresponding {@link QueueProcessor} for each of them
+     * until all queues have been processed.
+     *
+     * @throws GarbageCollectorException if anything goes wrong while processing the cleanup jobs
+     */
+    private void processCleanupJobs() throws GarbageCollectorException {
+        for(Map.Entry<Class<? extends GarbageCollectorJob>, ArrayDeque<GarbageCollectorJob>> entry : garbageCollectorJobs.entrySet()) {
+            if(shuttingDown) {
+                return;
+            }
+
+            QueueProcessor queueProcessor = queueProcessors.get(entry.getKey());
+            if(queueProcessor == null) {
+                throw new GarbageCollectorException("could not determine a queue processor for cleanup job of type " + entry.getKey().getCanonicalName());
+            }
+
+            queueProcessor.processQueue(this, entry.getValue());
+        }
+    }
+
+    /**
+     * This method retrieves the job queue belonging to a given job type.
+     *
+     * It first checks if a corresponding queue exists already and creates a new one if no queue was created yet for the
+     * given job type.
+     *
+     * @param jobClass type of the job that we want to retrieve the queue for
+     * @return the list of jobs for the provided job type
+     */
+    private ArrayDeque<GarbageCollectorJob> getJobQueue(Class<? extends GarbageCollectorJob> jobClass) {
+        if (garbageCollectorJobs.get(jobClass) == null) {
+            synchronized(this) {
+                if (garbageCollectorJobs.get(jobClass) == null) {
+                    garbageCollectorJobs.put(jobClass, new ArrayDeque<>());
```

Comment:
This should be blocking (it's single threaded right now anyway) just so a future parallelisation of this doesn't cause unexpected races.

Reply:
It's single threaded by design but what do you mean with concurrent? Instead of the ArrayDeque? because the getter is already thread safe by using double-checked locking.

Anyway here the reasons which i already posted in slack for doing this sequentially / single threaded:

1. this is mostly IO bound so spawning additional threads would not increase the througput alot - in fact it could even be harmfull - let me comment stackoverflow "IO to a single physical disk is intrinsically a serialized operation. Each thread would have to block, waiting for its chance to get access to the disk. In this case, multiple threads are probably useless...and may even lead to contention problems."

2. this is pretty much the least priority task in the node itself - we do not want to affect the overall performance of a node just to speed up the pruning (which due to 1. will not scale very well with the amount of workers anyway).

3. it allows us to compact the queues alot reducing both the memory footprint and the required hdd space to persist the the queues (in a secnario where a big db file needs to be cleaned up) we only need 48bytes to represent the pruning state (compared to possibly hundreds or even gigabytes of data if we need to be able to randomly address any job in the queue to take note that we pruned him up already - at some point persisting the state (to be able to continue) will require much more resources then the pruning itself)

4. since we need to do deletions atomicly in transactions because we otherwise loose the ability to address certain transactions if their connections to the tangle have been removed already and they are floating now) having multiple threads would also not be a good thing since it would increase the time it takes for 1 sub task to finish and therefore give IRI more chances to crash in between

5. since the addition of jobs gets triggered by the snapshot taking routine - which needs to lock the snapshot state while it is taking the snapshot we want additions to happen as fast as possible
so we can unlock the snapshots (which are the representation of the ledger state) as fast as possible without blocking the rest of the node
so having scenarios where we need to synchronize the processing and changes of the queues would reduce the performance of snapshots being taken and therefore the performance of the whole node